File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/multitrack/viewmodel/MultiTrackViewModel.kt
package com.musicmuni.voxatrace.demo.sections.multitrack.viewmodel

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.musicmuni.voxatrace.sonix.SonixMixer
import com.musicmuni.voxatrace.sonix.SonixMixerConfig
import io.github.aakira.napier.Napier
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

/**
 * ViewModel for multi-track mixing using SonixMixer.
 */
class MultiTrackViewModel : ViewModel() {

    // MARK: - Published State

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> = _isPlaying.asStateFlow()

    private val _currentTimeMs = MutableStateFlow(0L)
    val currentTimeMs: StateFlow<Long> = _currentTimeMs.asStateFlow()

    private val _durationMs = MutableStateFlow(0L)
    val durationMs: StateFlow<Long> = _durationMs.asStateFlow()

    private val _backingVolume = MutableStateFlow(0.8f)
    val backingVolume: StateFlow<Float> = _backingVolume.asStateFlow()

    private val _vocalVolume = MutableStateFlow(1.0f)
    val vocalVolume: StateFlow<Float> = _vocalVolume.asStateFlow()

    private val _status = MutableStateFlow("Ready")
    val status: StateFlow<String> = _status.asStateFlow()

    private val _isLoaded = MutableStateFlow(false)
    val isLoaded: StateFlow<Boolean> = _isLoaded.asStateFlow()

    // MARK: - Private State

    private var mixer: SonixMixer? = null
    private val observerJobs = mutableListOf<Job>()

    // MARK: - Actions

    fun initialize(context: Context) {
        if (_isLoaded.value) return

        viewModelScope.launch {
            try {
                _status.value = "Loading tracks..."

                // Copy assets to files
                val (backingPath, vocalPath) = withContext(Dispatchers.IO) {
                    val backingFile = copyAssetToFile(context, "sample.m4a")
                    val vocalFile = copyAssetToFile(context, "vocal.m4a")
                    Pair(backingFile.absolutePath, vocalFile.absolutePath)
                }

                // Create mixer with Config + Factory pattern (ADR-001)
                val mixerConfig = SonixMixerConfig.Builder()
                    .loopCount(1)
                    .onPlaybackComplete {
                        Napier.d("Playback complete!")
                        _status.value = "Playback complete"
                    }
                    .onLoopComplete { loopIndex ->
                        Napier.d("Completed loop $loopIndex")
                    }
                    .onError { error ->
                        Napier.e("Playback error: $error")
                        _status.value = "Error: $error"
                    }
                    .build()
                val newMixer = SonixMixer.create(mixerConfig)

                mixer = newMixer

                // Add tracks
                val backingSuccess = newMixer.addTrack("backing", backingPath)
                val vocalSuccess = newMixer.addTrack("vocal", vocalPath)

                if (backingSuccess && vocalSuccess) {
                    _durationMs.value = newMixer.duration
                    _isLoaded.value = true
                    _status.value = "Loaded 2 tracks"
                    Napier.d("Multi-track loaded: duration=${newMixer.duration}ms")

                    setupObservers(newMixer)
                } else {
                    _status.value = "Failed to load tracks"
                }
            } catch (e: Exception) {
                Napier.e("MultiTrack init failed", e)
                _status.value = "Error: ${e.message}"
            }
        }
    }

    private fun setupObservers(mixer: SonixMixer) {
        observerJobs.add(viewModelScope.launch {
            mixer.isPlaying.collect { playing ->
                _isPlaying.value = playing
            }
        })

        observerJobs.add(viewModelScope.launch {
            mixer.currentTime.collect { time ->
                _currentTimeMs.value = time
            }
        })

        observerJobs.add(viewModelScope.launch {
            mixer.error.collect { error ->
                error?.let {
                    _status.value = "Error: ${it.message}"
                }
            }
        })
    }

    fun play() {
        mixer?.play()
    }

    fun pause() {
        mixer?.pause()
    }

    fun stop() {
        mixer?.stop()
        _currentTimeMs.value = 0
    }

    fun seek(fraction: Float) {
        val seekPos = (fraction * _durationMs.value).toLong()
        mixer?.seek(seekPos)
    }

    fun setBackingVolume(volume: Float) {
        _backingVolume.value = volume
        mixer?.setTrackVolume("backing", volume)
    }

    fun setVocalVolume(volume: Float) {
        _vocalVolume.value = volume
        mixer?.setTrackVolume("vocal", volume)
    }

    fun fadeVocalDown() {
        mixer?.fadeTrackVolume("vocal", 0.2f, 500)
        _vocalVolume.value = 0.2f
    }

    fun fadeVocalUp() {
        mixer?.fadeTrackVolume("vocal", 1.0f, 500)
        _vocalVolume.value = 1.0f
    }

    override fun onCleared() {
        super.onCleared()
        observerJobs.forEach { it.cancel() }
        mixer?.release()
    }

    private fun copyAssetToFile(context: Context, assetName: String): File {
        val file = File(context.cacheDir, assetName)
        if (!file.exists()) {
            context.assets.open(assetName).use { input ->
                FileOutputStream(file).use { output ->
                    input.copyTo(output)
                }
            }
        }
        return file
    }

    companion object {
        fun formatTime(ms: Long): String {
            val seconds = (ms / 1000) % 60
            val minutes = (ms / 1000) / 60
            return String.format("%d:%02d", minutes, seconds)
        }
    }
}

File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/MultiTrack/MultiTrackViewModel.swift
import Foundation
import Combine
import VoxaTrace

/// ViewModel for multi-track mixing using SonixMixer.
///
/// ## VoxaTrace Integration (~20 lines)
/// ```swift
/// // 1. Create mixer with Config + Factory pattern
/// let config = SonixMixerConfig.Builder()
///     .loopCount(1)
///     .onPlaybackComplete { }
///     .onLoopComplete { }
///     .onError { }
///     .build()
/// mixer = SonixMixer.create(config: config, audioSession: .playback)
///
/// // 2. Add tracks (auto-decodes files)
/// try await mixer.addTrack(name: "backing", filePath: path)
///
/// // 3. Observe state
/// mixer.observeIsPlaying { }
/// mixer.observeCurrentTime { }
///
/// // 4. Control
/// mixer.play() / mixer.pause() / mixer.stop()
/// mixer.setTrackVolume(name:, volume:)
/// mixer.fadeTrackVolume(name:, targetVolume:, durationMs:)
/// ```
@MainActor
final class MultiTrackViewModel: ObservableObject {

    // MARK: - Published State

    @Published private(set) var isPlaying = false
    @Published private(set) var currentTimeMs: Int64 = 0
    @Published private(set) var durationMs: Int64 = 0
    @Published private(set) var isLoaded = false
    @Published private(set) var status = "Ready"
    @Published var backingVolume: Float = 0.8
    @Published var vocalVolume: Float = 1.0

    // MARK: - Computed Properties

    var seekProgress: Double {
        durationMs > 0 ? Double(currentTimeMs) / Double(durationMs) : 0
    }

    var formattedCurrentTime: String { formatTime(currentTimeMs) }
    var formattedDuration: String { formatTime(durationMs) }

    // MARK: - Private

    private var mixer: SonixMixer?
    private var playingTask: Task<Void, Never>?
    private var timeTask: Task<Void, Never>?
    private var errorTask: Task<Void, Never>?

    // MARK: - Lifecycle

    func onAppear() async {
        await initializeMixer()
    }

    func onDisappear() {
        cancelObservers()
        mixer?.release()
    }

    // MARK: - Actions

    func play() {
        mixer?.play()
    }

    func pause() {
        mixer?.pause()
    }

    func stop() {
        mixer?.stop()
        currentTimeMs = 0
    }

    func seek(to progress: Double) {
        let seekPos = Int64(progress * Double(durationMs))
        mixer?.seek(positionMs: seekPos)
    }

    func setBackingVolume(_ volume: Float) {
        backingVolume = volume
        mixer?.setTrackVolume(name: "backing", volume: volume)
    }

    func setVocalVolume(_ volume: Float) {
        vocalVolume = volume
        mixer?.setTrackVolume(name: "vocal", volume: volume)
    }

    func fadeVocalDown() {
        mixer?.fadeTrackVolume(name: "vocal", targetVolume: 0.2, durationMs: 500)
        vocalVolume = 0.2
    }

    func fadeVocalUp() {
        mixer?.fadeTrackVolume(name: "vocal", targetVolume: 1.0, durationMs: 500)
        vocalVolume = 1.0
    }

    // MARK: - Private Methods

    private func initializeMixer() async {
        status = "Loading tracks..."

        guard let backingPath = copyAssetToFile(name: "sample", ext: "m4a"),
              let vocalPath = copyAssetToFile(name: "vocal", ext: "m4a") else {
            status = "Assets not found"
            return
        }

        // Create mixer with Config + Factory pattern
        let mixerConfig = SonixMixerConfig.Builder()
            .loopCount(1)
            .onPlaybackComplete { [weak self] in
                Log.i(.mixer, "Playback complete")
                Task { @MainActor in
                    self?.status = "Playback complete"
                }
            }
            .onLoopComplete { loopIndex in
                Log.i(.mixer, "Loop \(loopIndex) completed")
            }
            .onError { [weak self] error in
                Log.e(.mixer, "Playback error: \(error)")
                Task { @MainActor in
                    self?.status = "Error: \(error)"
                }
            }
            .build()
        let newMixer = SonixMixer.create(config: mixerConfig, audioSession: .playback)

        // Auto-decode tracks
        let backingSuccess = try? await newMixer.addTrack(name: "backing", filePath: backingPath)
        let vocalSuccess = try? await newMixer.addTrack(name: "vocal", filePath: vocalPath)

        mixer = newMixer

        if backingSuccess == true && vocalSuccess == true {
            durationMs = newMixer.duration
            isLoaded = true
            status = "Loaded 2 tracks"
            Log.i(.mixer, "Multi-track loaded: duration=\(newMixer.duration)ms, tracks=2")

            setupObservers(newMixer)
        } else {
            Log.e(.mixer, "Failed to load tracks: backing=\(backingSuccess ?? false), vocal=\(vocalSuccess ?? false)")
            status = "Failed to load tracks"
        }
    }

    private func setupObservers(_ mixer: SonixMixer) {
        playingTask = mixer.observeIsPlaying { [weak self] playing in
            self?.isPlaying = playing
        }
        timeTask = mixer.observeCurrentTime { [weak self] time in
            self?.currentTimeMs = time
        }
        errorTask = mixer.observeError { [weak self] error in
            if let error = error {
                self?.status = "Error: \(error.message)"
            }
        }
    }

    private func cancelObservers() {
        playingTask?.cancel()
        timeTask?.cancel()
        errorTask?.cancel()
    }

    private func formatTime(_ ms: Int64) -> String {
        let seconds = (ms / 1000) % 60
        let minutes = (ms / 1000) / 60
        return String(format: "%d:%02d", minutes, seconds)
    }
}
